### ISSUE TITLE
Crafting recipes on machines not being marked as completed when results are fed into another recipe

### ISSUE THREAD

**Issue by cleavera** [CyclopsMC/IntegratedCrafting#48]
#### Issue type:

- :bug: Bug 

____

#### Short description:

Crafting recipes which are immediately fed into another crafting recipe are not always marked as completed.

#### Steps to reproduce the problem:

![Setup](https://anthonycleavercom.blob.core.windows.net/bush/debug.png "Setup")
This example uses the recipes: Cobblestone -> Gravel -> Sand. But any recipe feeding into another one works
1. Set up two recipes one for Cobblestone -> Gravel and one for Gravel -> Sand.
2. Place down two pulverisers, attach crafting interfaces to both of them and place your crafting recipes in them.
3. Place two crafting writers and give one a variable card for the gravel item and the other sand. Configure them to ignore inventory.
4. Attach a crafting job terminal to the network to view the progress of the crafting, place a storage container on an item interface to give the system some capacity
5. Feed system with a bunch of cobble
6. After about 10-15 iterations the gravel recipe will get stuck on crafting and never complete

#### Expected behaviour:

Marking the recipe as complete should happen before the ingredients are exported to the next crafting process.

____

#### Versions:

- This mod: IntegratedCrafting-1.12.2-1.0.10.jar
- Minecraft: 1.12.2
- Forge: 14.23.5.2854

#### Log file:

N/A

----
Alternative reconstruction setup using logs * 9 -> charcoal * 9 -> charcoal block
![Setup](https://anthonycleavercom.blob.core.windows.net/bush/debug2.png "Setup")
The easiest way to get this to consistently break is to have your `logs -> charcoal` recipe as `Logs * 9 -> Charcoal * 9`

**Comment by rubensworks** [comment]
Thanks for reporting!



<div align="right"><a href="https://github.com/CyclopsMC/IntegratedCrafting/actions/runs/383441608">Log</a> | <a href="https://github.com/peaceiris/actions-label-commenter#readme">Bot Usage</a></div>

**Comment by rubensworks** [comment]
I have not been able to reproduce this problem myself unfortunately.

Would it be possible to share the world of a reproducible setup? (preferably without any other mods, so I can load it into my dev environment for easier debugging)

**Comment by cleavera** [comment]
@rubensworks 

Apologies for the slow response, I must have missed the notification in the glut of Christmas emails. I wasn't able to recreate with a completely blank world, I ended up adding the furnace overhaul mod. It's unclear whether I could have recreated it with a vanilla furnace given enough time but with the zenith furnace it breaks fairly quickly.

https://anthonycleavercom.blob.core.windows.net/bush/debug.zip

I added a separate system to dispose of the glass panes once they're crafted as the chest kept clogging up with items. I'm not sure if the issue might be affected by lag as I was able to reproduce it a lot easier in my modpack testing world which has a bit of a lag issue.

<blockquote></blockquote>

**Comment by 7eggert** [comment]
*UNTESTED*

I noticed that IC does expect the results to show up in the storage. Thus your recipe needs to not mention the intermediate result but the final result.
E.g. you pulverize iron ore and then melt it; your recipe needs to say "Iron ore turns into two iron ingot".

**Comment by rubensworks** [comment]
I postponed this issue a bit due to its complexity...
But after many hours of debugging, I finally found the cause of it and fixed it!
It was due to a nasty race condition (feel free to read the commit messages if you want to know more).

I won't make new MC 1.12 releases, so I intend to port this change to 1.16 and release the fix there.

@cleavera Before I release these fixes, would you be willing to confirm this change also fixes everything on your end (in 1.12)?
Since it's possible that you're not tracking this issue anymore, I will consider this issue fixed if I don't get a reply by Monday.

Dev builds can be found here:
* https://www.dropbox.com/s/mbdt8n020vvcj7f/IntegratedCrafting-1.12.2-1.0.10-DEV.jar?dl=0
* https://www.dropbox.com/s/xrn3egeoeqmznng/IntegratedDynamics-1.12.2-1.1.11-DEV.jar?dl=0

<blockquote><img src="https://www.dropbox.com/static/images/spectrum-icons/generated/content/content-unknown-large.png" width="48" align="right"><div>Dropbox</div><div><strong><a href="https://www.dropbox.com/s/mbdt8n020vvcj7f/IntegratedCrafting-1.12.2-1.0.10-DEV.jar?dl=0">IntegratedCrafting-1.12.2-1.0.10-DEV.jar</a></strong></div><div>Shared with Dropbox</div></blockquote>
<blockquote><img src="https://www.dropbox.com/static/images/spectrum-icons/generated/content/content-unknown-large.png" width="48" align="right"><div>Dropbox</div><div><strong><a href="https://www.dropbox.com/s/xrn3egeoeqmznng/IntegratedDynamics-1.12.2-1.1.11-DEV.jar?dl=0">IntegratedDynamics-1.12.2-1.1.11-DEV.jar</a></strong></div><div>Shared with Dropbox</div></blockquote>

Also, I disabled the glass pane trash for testing, as it caused some conflicts due to different ID networks not communicating with each other, so it's better if you'd do this as well.
